src/multivariate.jl

```diff
@@ -1,52 +1,82 @@
+end
+Base.length(d::TuringDenseMvNormal) = length(d.m)
+function Distributions.rand(rng::Random.AbstractRNG, d::TuringDenseMvNormal)
     return d.m .+ d.C.U' * randn(rng, dim(d))
 end
+function Distributions.rand(rng::Random.AbstractRNG, d::TuringDenseMvNormal, n::Int)
+    return d.m .+ d.C.U' * randn(rng, dim(d), n)
+end
 
 """
-    TuringDiagNormal{Tm<:AbstractVector, Tσ<:AbstractVector} <: ContinuousMultivariateDistribution
+    TuringDiagMvNormal{Tm<:AbstractVector, Tσ<:AbstractVector} <: ContinuousMultivariateDistribution
 
 A multivariate normal distribution whose covariance is diagonal. Compatible with Tracker.
 """
-struct TuringDiagNormal{Tm<:AbstractVector, Tσ<:AbstractVector} <: ContinuousMultivariateDistribution
+struct TuringDiagMvNormal{Tm<:AbstractVector, Tσ<:AbstractVector} <: ContinuousMultivariateDistribution
     m::Tm
     σ::Tσ
 end
 
-Distributions.dim(d::TuringDiagNormal) = length(d.m)
-function Distributions.rand(rng::Random.AbstractRNG, d::TuringDiagNormal)
-    return d.m .+ d.σ .* randn(rng, dim(d))
+Base.length(d::TuringDiagMvNormal) = length(d.m)
+Base.size(d::TuringDiagMvNormal) = (length(d), length(d))
+function Distributions.rand(rng::Random.AbstractRNG, d::TuringDiagMvNormal)
+    return d.m .+ d.σ .* randn(rng, length(d))
 end
+function Distributions.rand(rng::Random.AbstractRNG, d::TuringDiagMvNormal, n::Int)
+    return d.m .+ d.σ .* randn(rng, length(d), n)
+end
+
+struct TuringScalMvNormal{Tm<:AbstractVector, Tσ<:Real} <: ContinuousMultivariateDistribution
+    m::Tm
+    σ::Tσ
+end
+
+Base.length(d::TuringScalMvNormal) = length(d.m)
+Base.size(d::TuringScalMvNormal) = (length(d), length(d))
+function Distributions.rand(rng::Random.AbstractRNG, d::TuringScalMvNormal)
+    return d.m .+ d.σ .* randn(rng, length(d))
+end
+function Distributions.rand(rng::Random.AbstractRNG, d::TuringScalMvNormal, n::Int)
+    return d.m .+ d.σ .* randn(rng, length(d), n)
+end
+
 for T in (:AbstractVector, :AbstractMatrix)
-    @eval Distributions.logpdf(d::TuringDiagNormal, x::$T) = _logpdf(d, x)
-    @eval Distributions.logpdf(d::TuringMvNormal, x::$T) = _logpdf(d, x)
+    @eval Distributions.logpdf(d::TuringScalMvNormal, x::$T) = _logpdf(d, x)
+    @eval Distributions.logpdf(d::TuringDiagMvNormal, x::$T) = _logpdf(d, x)
+    @eval Distributions.logpdf(d::TuringDenseMvNormal, x::$T) = _logpdf(d, x)
 end
 
-function _logpdf(d::TuringDiagNormal, x::AbstractVector)
-    return -(dim(d) * log(2π) + 2 * sum(log.(d.σ)) + sum(abs2, (x .- d.m) ./ d.σ)) / 2
+function _logpdf(d::TuringScalMvNormal, x::AbstractVector)
+    return -(length(x) * log(2π * abs2(d.σ)) + sum(abs2, (x .- d.m) ./ d.σ)) / 2
 end
-function _logpdf(d::TuringDiagNormal, x::AbstractMatrix)
-    return -(dim(d) * log(2π) .+ 2 * sum(log.(d.σ)) .+ sum(abs2, (x .- d.m) ./ d.σ, dims=1)') ./ 2
+function _logpdf(d::TuringScalMvNormal, x::AbstractMatrix)
+    return -(size(x, 2) * log(2π) .+ 2 * sum(log(d.σ)) .+ sum(abs2, (x .- d.m) ./ d.σ, dims=1)') ./ 2
```

Comment:
This seems wrong, I think it should be
```suggestion
    return -(size(x, 2) * (log(2π) + 2 * log(d.σ)) .+ sum(abs2, (x .- d.m) ./ d.σ, dims=1)') ./ 2
```
or
```suggestion
    return -(size(x, 2) * log(2π * abs2(d.σ)) .+ sum(abs2, (x .- d.m) ./ d.σ, dims=1)') ./ 2
```
as in the vector-valued case above. IMO it's a bit surprising to use `size(x, 2)` and `length(x)` instead of `length(d)`, but it doesn't matter I guess.

I'm also wondering if you actually want to use `sum(abs2, ....)'` since that will just transpose the vector but keep both dimensions
```julia
julia> sum(abs2, rand(10, 20); dims = 1)'
20×1 LinearAlgebra.Adjoint{Float64,Array{Float64,2}}:
```
Hence the result is not a `Vector{<:Real}` even though the default implementation in Distributions returns a vector:

https://github.com/JuliaStats/Distributions.jl/blob/70f3d97fca49b0abfa154f2f29c7b4f4a90b2ff8/src/multivariates.jl#L230-L235

Reply:
I guess you want to use
```suggestion
    return -(size(x, 2) * (log(2π) + 2 * log(d.σ)) .+ vec(sum(abs2, (x .- d.m) ./ d.σ, dims=1))) ./ 2
```

Reply:
Or maybe even
```suggestion
    return -(size(x, 2) * (log(2π) + 2 * log(d.σ)) .+ vec(sum(abs2, x .- d.m, dims=1)) ./ abs2(d.σ)) ./ 2
```

Reply:
Good catch. Thanks. I will fix and test the matrix case.